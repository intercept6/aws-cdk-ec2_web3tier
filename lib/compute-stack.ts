import cdk = require('@aws-cdk/core');
import {AutoScalingGroup} from "@aws-cdk/aws-autoscaling";
import {
    AmazonLinuxImage,
    InstanceClass,
    InstanceSize,
    InstanceType,
    ISecurityGroup,
    IVpc,
    SubnetType
} from "@aws-cdk/aws-ec2";
import {ApplicationLoadBalancer} from "@aws-cdk/aws-elasticloadbalancingv2";
import {Role} from "@aws-cdk/aws-iam";

interface ComputeStackProps extends cdk.StackProps {
    vpc: IVpc;
    instanceRole: Role;
    appSg: ISecurityGroup;
}

export class ComputeStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props: ComputeStackProps) {
        super(scope, id, props);


        const prj: string = this.node.tryGetContext('prj');
        const stage: string = this.node.tryGetContext('stage');
        // const params: any = this.node.tryGetContext(stage);

        const asg = new AutoScalingGroup(this, 'ASG', {
            vpc: props.vpc,
            instanceType: InstanceType.of(InstanceClass.BURSTABLE2, InstanceSize.MICRO),
            machineImage: new AmazonLinuxImage(),
            allowAllOutbound: true,
            role: props.instanceRole
        });
        asg.addUserData(
            "yum -y update",
            "yum -y install nginx",
            "systemctl enable nginx",
            "systemctl start nginx"
        );

        const alb = new ApplicationLoadBalancer(this, 'Alb', {
            vpc: props.vpc,
            vpcSubnets: {subnetType: SubnetType.PUBLIC},
            internetFacing: true,
            loadBalancerName: `${prj}-${stage}-alb`
        });

        const listener = alb.addListener('Listener', {
            port: 80,
        });
        listener.addTargets('Target', {
            targets: [asg],
            port: 80
        });

    }
}