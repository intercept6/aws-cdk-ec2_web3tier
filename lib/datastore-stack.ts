// @aws-cdk/aws-rds is a developer preview (public beta) module.
// Releases might lack important features and might have future breaking changes.

import {
  Construct,
  RemovalPolicy,
  Stack,
  StackProps,
  Tag
} from "@aws-cdk/core";
import {
  InstanceClass,
  InstanceSize,
  InstanceType,
  ISecurityGroup,
  IVpc,
  SubnetType
} from "@aws-cdk/aws-ec2";
import { DatabaseCluster, DatabaseClusterEngine } from "@aws-cdk/aws-rds";

interface DataStoreStackProps extends StackProps {
  vpc: IVpc;
  dbSg: ISecurityGroup;
}

export class DataStoreStack extends Stack {
  constructor(scope: Construct, id: string, props: DataStoreStackProps) {
    super(scope, id, props);

    const prj: string = this.node.tryGetContext("prj");
    const stage: string = this.node.tryGetContext("stage");
    // const params: any = this.node.tryGetContext(stage);

    const cluster = new DatabaseCluster(this, "DBCluster", {
      engine: DatabaseClusterEngine.AURORA,
      masterUser: {
        username: "admin"
      },
      removalPolicy: RemovalPolicy.DESTROY,
      instanceProps: {
        instanceType: InstanceType.of(
          InstanceClass.MEMORY5,
          InstanceSize.LARGE
        ),
        vpcSubnets: {
          subnetType: SubnetType.ISOLATED
        },
        vpc: props.vpc,
        securityGroup: props.dbSg
      }
    });
    cluster.node.applyAspect(new Tag("Name", `${prj}-${stage}-db`));
  }
}
