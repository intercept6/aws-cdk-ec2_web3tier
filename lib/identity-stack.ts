import { Construct, Stack, StackProps } from "@aws-cdk/core";
import { ManagedPolicy, Role, ServicePrincipal } from "@aws-cdk/aws-iam";

export class IdentityStack extends Stack {
  public readonly instanceRole: Role;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const prj: string = this.node.tryGetContext("prj");
    const stage: string = this.node.tryGetContext("stage");
    // const params: object = this.node.tryGetContext(stage);

    this.instanceRole = new Role(this, "IamRole", {
      assumedBy: new ServicePrincipal("ec2.amazonaws.com"),
      roleName: `${prj}-${stage}-app`,
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AmazonEC2RoleforSSM"
        )
      ]
    });
  }
}
