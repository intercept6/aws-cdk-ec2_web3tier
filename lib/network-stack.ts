import { Construct, Stack, StackProps, Tag } from "@aws-cdk/core";
import { Port, SecurityGroup, SubnetType, Vpc } from "@aws-cdk/aws-ec2";

export class NetworkStack extends Stack {
    public readonly vpc: Vpc;
    public readonly appSg: SecurityGroup;
    public readonly dbSg: SecurityGroup;

    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const prj: string = this.node.tryGetContext("prj");
        const stage: string = this.node.tryGetContext("stage");
        const params: any = this.node.tryGetContext(stage);

    this.vpc = new Vpc(this, "VPC", {
      enableDnsHostnames: true,
      enableDnsSupport: true,
      natGateways: params.nat_gateways,
      maxAzs: 2,
      subnetConfiguration: [
        { name: "Public", subnetType: SubnetType.PUBLIC, cidrMask: 24 },
        { name: "Private", subnetType: SubnetType.PRIVATE, cidrMask: 24 },
        { name: "Isolated", subnetType: SubnetType.ISOLATED, cidrMask: 24 }
      ]
    });
    this.vpc.node.applyAspect(new Tag("Name", `${prj}-${stage}-vpc`));

    this.appSg = new SecurityGroup(this, "AppSg", {
      allowAllOutbound: true,
      vpc: this.vpc,
      securityGroupName: `${prj}-${stage}-app`,
      description: `${prj}-${stage}-app`
    });
    this.appSg.addIngressRule(
      this.appSg,
      Port.tcp(80),
      "allow http access from alb"
    );
    this.appSg.node.applyAspect(new Tag("Name", `${prj}-${stage}-app`));

    this.dbSg = new SecurityGroup(this, "DbSg", {
      allowAllOutbound: true,
      vpc: this.vpc,
      securityGroupName: `${prj}-${stage}-db`,
      description: `${prj}-${stage}-db`
    });
    this.dbSg.addIngressRule(
      this.appSg,
      Port.tcp(3306),
      "allow db access from app server"
    );
    this.dbSg.node.applyAspect(new Tag("Name", `${prj}-${stage}-db`));
  }
}
