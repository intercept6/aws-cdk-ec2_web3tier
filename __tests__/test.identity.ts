import { App } from "@aws-cdk/core";
// import { InstanceType, UserData, Vpc } from "@aws-cdk/aws-ec2"
import { SynthUtils } from "@aws-cdk/assert";
import { IdentityStack } from "../lib/identity-stack";

describe("identity", () => {
  test("default", () => {
    const app = new App();
    app.node.setContext("prj", "ssprj");
    app.node.setContext("stage", "ssstage");
    const stack = new IdentityStack(app, "TestIdentityStack");
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
