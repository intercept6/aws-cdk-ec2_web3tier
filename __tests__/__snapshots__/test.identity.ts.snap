// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`identity default 1`] = `
Object {
  "Resources": Object {
    "IamRoleA750FF82": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonEC2RoleforSSM",
              ],
            ],
          },
        ],
        "RoleName": "ssprj-ssstage-app",
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
